design/cluster-api-provider-metal3/configurable_disk_cleaning.md

```diff
@@ -0,0 +1,260 @@
+
+- Enable disabling automated cleaning for 1/N nodes in Ironic.
+
+### Non-Goals
+
+None
+
+## Proposal
+
+The implementation of this proposal needs to be done both in Metal³
+and Ironic. Metal³ needs to be able to identify the nodes for which the
+disk cleaning should be disabled/enabled and mark them in a way that Ironic
+would understand and avoid cleaning for those specific nodes.
+
+### Ironic
+
+Ironic supports enabling/disabling automated cleaning globally. It also
+allows to enable automated cleaning for a node. However, once enabled,
+there is no way to disable it again, because currently node.automated_clean
+respects only `None` or `True` values. As such, first the
+[2008113](https://storyboard.openstack.org/#!/story/2008113) needs to be
+implemented. Once the Ironic supports disabling of disk cleaning per node,
+Baremetal Operator can start requesting disabling/enabling disk cleaning.
+
+### Metal³
+
+In the Metal³, we will need changes both in Cluster-api-provider-metal3(CAPM3)
+and Baremetal Operator(BMO) as follows.
+
+#### Cluster-api-provider-metal3
+
+##### Metal3MachineTemplate object
+
+We introduce a new field `disableAutomatedClean` in Metal3MachineTemplate.
+which stores a boolean value. Setting it to False means don’t perform
+automated disk cleaning, while true value enables automated disk cleaning.
+By default disableAutomatedClean will be set to False (i.e. do perform automated
+cleaning).
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: Metal3MachineTemplate
+metadata:
+spec:
+  template:
+  spec:
+   disableAutomatedClean: True
+```
+
+A new Metal3MachineTemplate controller will be introduced to reconcile
+Metal3MachineTemplate objects. When a user modifies `disableAutomatedClean`
+field on a Metal3MachineTemplate, CAPM3 Metal3MachineTemplate controller
+will catch it and update `disableAutomatedClean` field value on all the
+Metal3Machines referenced by that Metal3MachineTemplate.
+
+##### Metal3Machine object
+
+We introduce a new field `disableAutomatedClean` in Metal3Machine which stores
+a boolean value. Setting it to false means don’t perform automated disk cleaning,
+while true value enables automated disk cleaning. By default `disableAutomatedClean`
+will be set to False (i.e. do perform automated cleaning).
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: Metal3Machine
+metadata:
+spec:
+  template:
+  spec:
+   disableAutomatedClean: True
+
+```
+
+##### Metal3Machine Controller
+
+Currently Metal3Machine controller removes the consumerRef on a BaremetalHost
+when it goes to deprovisioning. However, we are proposing to set the name of
+KubeadmControlPlane(KCP)/MachineDeployment(MD) in the consumerRef when
+BaremetalHost is going for deprovisioning.This will avoid the BaremetalHost
+to be utilized by any other Metal3Machines apart from the ones that were used
+before starting the upgrade.
```

Comment:
This is actually unrelated to the configurable disk cleaning. It is needed to achieve predictability of the hosts selected during the process (and the main reason why we are implementing this disk cleaning configurability) but not strictly linked to it. Could you move it somewhere else in this proposal or even in another proposal centered on node re-use during upgrades ?

Reply:
Yes I removed node reuse part from this proposal, and will open a new one for that part. The current one includes only disk cleaning.